### PR TITLE
docs: add dubbo-go agent skills

### LINKED ISSUE #3334: [FEATURE] 已经支持AI Code Review 或者 什么AI check List了吗？
### ✅ 验证清单

- [x] 🔍 我已经搜索过 [现有 Issues](https://github.com/apache/dubbo-go/issues)，确信这不是重复请求
- [x] 📋 我已经查看了 [发布说明](https://github.com/apache/dubbo-go/releases)，确信此功能尚未实现

### 🎯 功能描述

是否支持AI优化的claude.md/Agents.md文件，以及多层级的文档。

### 📋 使用场景

AI Coding 场景，用于约束AI编程

### ⚖️ 复杂性与风险评估

_No response_

### 🔗 外部依赖

_No response_

### 📚 附加信息

_No response_

### PR DESCRIPTION
**Pull request by Oxidaner**
### Description
Fixes # (issue)

### Checklist
- [x] I confirm the target branch is `develop`
- [x] Code has passed local testing
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Alanxtl** [comment]
pls check this https://github.com/Tsukikage7/dubbo-go-skills which in under construction

**Comment by Alanxtl** [comment]
@Tsukikage7 check this

File: .agents/skills/dubbo-go-domain/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
+---
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-domain/SKILL.md]: [P0] 缺少 Apache 2.0 许可证头。Apache 项目要求所有源文件（含 Markdown 文档）包含标准许可证头。dubbo-go 项目 `doc/` 目录下的 Markdown 文件均有许可证头，`.agents/skills/` 目录下也不应例外。

建议：在文件顶部添加：
```
<!-- Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. ... -->
```
全部 19 个文件均需添加。

File: .agents/skills/dubbo-go-observability-filters/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+Return filter chain position, provider or consumer side, telemetry contract, changed files, and validation performed.
+
+## Validation
+
+- Confirm filter names are registered and selected through `service.filter` or `reference.filter`.
+- Confirm filters call the next invoker exactly once unless intentionally short-circuiting.
+- Confirm `OnResponse` handles both success and error results.
+- Run focused tests such as `go test ./filter/... ./metrics/... ./otel/...`; use `make test` for broad interception changes.
+
+## Edge cases
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-observability-filters/SKILL.md]: [P1] 路径引用有误：文档引用 `filter/otel/` 和 `filter/tracing/`，但 dubbo-go 实际代码中 OpenTelemetry 相关代码位于顶级 `otel/` 目录。需核实实际目录结构并修正。

File: .agents/skills/dubbo-go-domain/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+This repository does not own external registry servers, Dubbo Java implementation, sample applications outside this repository, or production deployment infrastructure.
+
+## Tech Stack
+
+- Language: Go, module `dubbo.apache.org/dubbo-go/v3`.
+- Go version: `go 1.25.0` in `go.mod`; root `Makefile` uses `GOTOOLCHAIN=go1.25.0+auto` for tests.
+- Build: Go modules plus nested Go modules under selected `tools/` directories.
+- Test: `go test`, `testify`, `gomock`, and focused package tests.
+- Lint and format: `go fmt`, `imports-formatter`, `go vet`, `golangci-lint`, and `make fmt` or `make lint`.
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-domain/SKILL.md]: [P1] Go 版本声明可疑：文档称 `go 1.25.0`，但截至 2026 年 5 月 Go 1.25 尚未正式发布（通常 8 月发布）。需核实 go.mod 实际内容，若 go.mod 确实写了 1.25.0 则注明这是 toolchain 指令而非最低版本要求。

File: .agents/skills/dubbo-go-protocol/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
+---
+name: dubbo-go-protocol
+description: >-
+  Implements and reviews dubbo-go protocol and transport changes. Use when the user asks about Protocol, Invoker,
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-protocol/SKILL.md]: [P1] 协议列表中包含 "HTTP/3 transport"，但 dubbo-go 主仓库是否包含 HTTP/3 实现需确认。HTTP/3 支持可能在独立仓库或分支中。若不存在则应删除此引用。

File: .agents/skills/dubbo-go-cluster/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
+---
+name: dubbo-go-cluster
+description: >-
+  Implements and reviews dubbo-go cluster, routing, directory, and load-balancing changes. Use when the user asks about
+  failover, failfast, failsafe, failback, forking, broadcast, available, zoneaware, adaptive service, Directory,
+  RouterChain, condition, tag, script, affinity or Polaris routers, load balance algorithms, retries, provider
+  selection, or packages cluster/cluster/, cluster/directory/, cluster/router/, and cluster/loadbalance/. Do not use
+  for registry backend mechanics unless the directory snapshot contract is affected.
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-cluster/SKILL.md]: [P1] cluster 策略列表可能不完整：列出 failover/failfast/failsafe/failback/forking/broadcast/available/zoneaware/adaptive，但需核实 `cluster/cluster/` 目录下所有策略实现。"adaptive" 在 dubbo-go 中通常与 service discovery 相关而非 cluster strategy，需确认。

File: .agents/skills/dubbo-go-domain/references/project-rules.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# dubbo-go Project Rules
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-domain/references/project-rules.md]: [P2] project-rules.md 与 domain SKILL.md 的 "Do Not" 和 "Coding Guidelines" 章节内容几乎完全重复。两处维护同一套规则会导致后续更新不一致。

建议：选择一处作为规范权威来源（建议 project-rules.md），SKILL.md 中仅引用而非复制。

File: .agents/skills/dubbo-go-runtime/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
+---
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-runtime/SKILL.md]: [P2] "runtime" 术语模糊：dubbo-go 代码库中没有 `runtime/` 目录，runtime 概念分散在 `client/`、`server/`、`proxy/`、`graceful_shutdown/` 中。SKILL 名称与代码组织不直接对应。

建议：考虑改名为 `dubbo-go-client-server` 或在文档中更明确地解释 "runtime" 指代哪些包。

File: .agents/skills/dubbo-go-tools/references/tooling-flow.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# Tooling Flow
```
Comment (AlexStocks) [inline .agents/skills/dubbo-go-tools/references/tooling-flow.md]: [P2] 工具流程图过于简略，是所有 reference 文档中最短的。缺少 protoc 插件工作流、schema 生成流程等关键信息。建议补充完整的工具链流程，特别是 `protoc-gen-go-triple` 的代码生成管线。

**Review by AlexStocks — CHANGES_REQUESTED** [review]
## Code Review — dubbo-go #3340

**不建议合并，需重大修改。**

**P0 × 1**：全部 19 个文件缺少 Apache 2.0 许可证头（Apache 项目硬性要求）。同时 PR 描述为模板化占位符 "Fixes # (issue)"，无具体 issue 关联和变更动机说明，违反 Apache 社区贡献规范。`.agents/` 目录命名（点号前缀）可能被 .gitignore 排除，需确认或改用 `agents/` 目录名。

**P1 × 4**：Go 版本号可疑（1.25.0）、目录路径引用有误（filter/otel/ → otel/）、协议列表不准确（HTTP/3）、cluster 策略列表不完整

**P2 × 3**：domain SKILL.md 与 project-rules.md 内容大量重复、runtime 术语与代码组织不对应、工具流程图过于简略

**改进优先级**：
1. 为全部 19 个文件添加 Apache 2.0 许可证头
2. 重写 PR 描述，关联 issue，说明变更动机
3. 确认 .agents/ 目录不被 .gitignore 排除
4. 逐项核实文档中的代码路径、API 签名、Go 版本号
5. 消除 domain SKILL.md 与 project-rules.md 的内容重复

**Comment by Oxidaner** [comment]
This pr is by the https://github.com/Tsukikage7/dubbo-go-skills version update. 
Main changes:
- Added a repository development skill: `development/SKILL.md`
- Refreshed the imported Tsukikage7 dubbo-go skills against the current `develop` branch
- Added Apache 2.0 license headers to commentable files under `.agents`

File: .agents/skills/dubbo-go-observability-filters/SKILL.md (hunk 2 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-observability-filters/SKILL.md]: I have refactored the code. Please reevaluate it

File: .agents/skills/dubbo-go-domain/SKILL.md (hunk 1 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-domain/SKILL.md]: done

File: .agents/skills/dubbo-go-domain/SKILL.md (hunk 3 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-domain/SKILL.md]: I have refactored the code. Please reevaluate it

File: .agents/skills/dubbo-go-protocol/SKILL.md (hunk 4 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-protocol/SKILL.md]: I have refactored the code. Please reevaluate it

File: .agents/skills/dubbo-go-cluster/SKILL.md (hunk 5 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-cluster/SKILL.md]: I have refactored the code. Please reevaluate it

File: .agents/skills/dubbo-go-domain/references/project-rules.md (hunk 6 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-domain/references/project-rules.md]: I have refactored the code. Please reevaluate it

File: .agents/skills/dubbo-go-runtime/SKILL.md (hunk 7 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-runtime/SKILL.md]: I have refactored the code. Please reevaluate it

File: .agents/skills/dubbo-go-tools/references/tooling-flow.md (hunk 8 above)
Comment (Oxidaner) [reply .agents/skills/dubbo-go-tools/references/tooling-flow.md]: I have refactored the code. Please reevaluate it

**Comment by Alanxtl** [comment]
可以加一下 参考文档在https://github.com/apache/dubbo-website/tree/master/content/zh-cn/overview/mannual/golang-sdk

**Comment by Oxidaner** [comment]
https://github.com/apache/dubbo-website/pull/3209

**Comment by Alanxtl** [comment]
> [apache/dubbo-website#3209](https://github.com/apache/dubbo-website/pull/3209)

我的意思是把有参考文档的这个事情写在skills里面

不过你这个pr加了就加了吧，也挺好的

**Review by Alanxtl — APPROVED** [review]

File: .agents/.codex/INSTALL.md
Comment (Alanxtl) [inline .agents/.codex/INSTALL.md:1]: 这个文档里面写的步骤都必要吗？我怎么感觉不太需要呀

File: .agents/skills/debug/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+**Cause**: Consumer cannot find a provider in the registry.
+
+Checklist:
+- [ ] Provider started successfully? Look for `dubbo server started` and `A provider service ... was registered successfully` in provider logs.
+- [ ] Same registry address on both sides (`dubbo.WithRegistry(registry.WithAddress(...))` / YAML `dubbo.registries.xxx.address`)?
+- [ ] Same application name on both sides (`dubbo.WithName(...)`)? v3 defaults to **application-level** discovery — the registry stores the app name, not the interface FQN.
+- [ ] Same `interface` name passed to `pb.RegisterXxxHandler` and `pb.NewXxxService`?
+- [ ] Same protocol on both sides (both `tri` or both `dubbo`)?
+- [ ] Provider visible in the registry?
```
Comment (Alanxtl) [inline .agents/skills/debug/SKILL.md:54]: 这里加一种情况：是否是同一等级的服务发现，dubbo-go最新版默认使用应用级服务发现，老版本默认使用接口级服务发现，这两者并不互通

File: .agents/skills/debug/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+```
+
+**Cause**: Provider and consumer using different serialization formats.
+
+Checklist:
+- [ ] Both sides on the same protocol (both `tri` or both `dubbo`)?
+- [ ] Protobuf: same `.proto` compiled on both sides? Same `go_package`?
+- [ ] Hessian2: POJO `JavaClassName()` matches the Java class FQN exactly? `RegisterPOJO` called in an `init()` that actually runs?
+
+For Hessian2 specifics, see `dubbo-go-java-interop`.
```
Comment (Alanxtl) [inline .agents/skills/debug/SKILL.md]: dubbo-go-java-interop加个链接

File: .agents/skills/debug/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+defer cancel()
+resp, err := svc.Greet(ctx, req)
+```
+
+## Filter not found / filter panic
+
+**Cause**: Filter imported but `init()` not registered, or name mismatch between `extension.SetFilter` and `WithFilter`.
+
+Checklist:
+- [ ] Blank import present? e.g. `_ "dubbo.apache.org/dubbo-go/v3/filter/token"` or `_ "github.com/yourorg/yourapp/filter/myfilter"`
```
Comment (Alanxtl) [inline .agents/skills/debug/SKILL.md]: 这里加一个例子`_ "github.com/apache/dubbo-go-extensions/filter/hystrix"`

File: .agents/skills/debug/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+if err := srv.Serve(); err != nil { panic(err) }
+
+// Wrong: non-blocking, process exits immediately
+go srv.Serve()
+```
+
+For graceful shutdown:
+```go
+import _ "dubbo.apache.org/dubbo-go/v3/graceful_shutdown"
+```
```
Comment (Alanxtl) [inline .agents/skills/debug/SKILL.md]: 现在有这个用法吗

File: .agents/skills/guide/samples-index.md
Comment (Alanxtl) [inline .agents/skills/guide/samples-index.md:1]: 直接让llm读dubbo-go-samples的readme不就完了，这个samples可能还会更新呢

File: .agents/.gitignore
Comment (Alanxtl) [inline .agents/.gitignore:1]: 不用单独弄一个gitingore，用根目录的不行吗

**Review by Alanxtl — CHANGES_REQUESTED** [review]

File: .agents/.codex/INSTALL.md
Comment (Oxidaner) [reply .agents/.codex/INSTALL.md:1]: 只保留Tell Codex: 剩下的删掉?

File: .agents/skills/debug/SKILL.md (hunk 9 above)
Comment (Oxidaner) [reply .agents/skills/debug/SKILL.md:54]: done

File: .agents/skills/debug/SKILL.md (hunk 10 above)
Comment (Oxidaner) [reply .agents/skills/debug/SKILL.md]: done

File: .agents/skills/debug/SKILL.md (hunk 11 above)
Comment (Oxidaner) [reply .agents/skills/debug/SKILL.md]: done

File: .agents/skills/guide/samples-index.md
Comment (Oxidaner) [reply .agents/skills/guide/samples-index.md:1]: done

File: .agents/skills/debug/SKILL.md (hunk 12 above)
Comment (Oxidaner) [reply .agents/skills/debug/SKILL.md]: done

File: .agents/.gitignore
Comment (Oxidaner) [reply .agents/.gitignore:1]: done

File: .agents/.codex/INSTALL.md
Comment (Alanxtl) [reply .agents/.codex/INSTALL.md:1]: 这个是面向用户的还是面向agents的文档呀

Comment (Oxidaner) [reply .agents/.codex/INSTALL.md:1]: 我晕了 I have fixed it.

Comment (Alanxtl) [reply .agents/.codex/INSTALL.md:1]: 我觉得无论是面向用户的还是面向agents这个文档都没啥用呀

**Comment by AlexStocks** [comment]
Gemini 删掉

File: .agents/.codex/INSTALL.md
Comment (Alanxtl) [reply .agents/.codex/INSTALL.md:1]: 你看一下https://github.com/obra/superpowers 就没有这个文档
 这个install.md 真的有必要吗

File: .agents/gemini-extension.json [hunk 13]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (AlexStocks) [inline .agents/gemini-extension.json]: [P1] 你在本 PR 里要求“Gemini 删掉”，但当前 head 里 .agents/GEMINI.md 和这份 .agents/gemini-extension.json 仍然作为新增文件存在（都被这个 PR 引入）。如果决定不带 Gemini 适配，需要在本 PR 一并 git rm 掉 GEMINI.md 和 gemini-extension.json，并从 package.json/plugin.json/marketplace.json 任何引用 gemini 的位置同步移除，避免后续 install 流程指向不存在的文件。

File: .agents/skills/scaffolding/SKILL.md [hunk 14]
```diff
@@ -0,0 +1,282 @@
+---
```
Comment (AlexStocks) [inline .agents/skills/scaffolding/SKILL.md:1]: [P1] 这一批 7 个 .agents/skills/*/SKILL.md 仍然没有 Apache 2.0 license header（debug、development、extensions、guide、java-interop、migrate、scaffolding 全部命中），与 PR 描述里“Added Apache 2.0 license headers to commentable files under .agents”不一致；同一批的 .codex/INSTALL.md、.opencode/plugins/dubbo-go-agent-skills.js、README*.md 都已加上，只剩 SKILL.md 漏了。Markdown 可以用 HTML 注释或 frontmatter 注释加 ASF header；这是 Apache 项目硬要求，需要在 7 个文件里全部补齐，否则 release audit 会卡。

File: .agents/.opencode/plugins/dubbo-go-agent-skills.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
```
Comment (AlexStocks) [inline .agents/.opencode/plugins/dubbo-go-agent-skills.js:8]: [P1] 新增插件文件未通过空白检查

`git diff --check` 在本文件第 8、10 行检测到尾随空格。该文件会作为插件入口分发，请先清除尾随空格并确保仓库格式检查通过。

File: .agents/.claude-plugin/marketplace.json [hunk 16]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (AlexStocks) [inline .agents/.claude-plugin/marketplace.json:1]: [P1] README 中的 Claude 安装入口无法从仓库根目录发现这个 marketplace
README 要求执行 `/plugin marketplace add apache/dubbo-go`，但 Claude 官方协议要求 GitHub marketplace 的 `.claude-plugin/marketplace.json` 位于仓库根目录；当前文件放在 `.agents/.claude-plugin/marketplace.json`。本地 `claude plugin validate .agents` 虽能通过，但对 README 实际指向的仓库根目录执行 `claude plugin validate .` 会稳定报 `No manifest found ... Expected .claude-plugin/marketplace.json or .claude-plugin/plugin.json`。此外 marketplace 的 source 指向整个仓库，而实际 skills/plugin root 位于 `.agents`。因此用户按文档操作时无法安装这些 skill。建议将 marketplace manifest 放到仓库根 `.claude-plugin/marketplace.json`，并用相对 `./.agents` 或 `git-subdir` 明确插件根；插件 manifest 也应按协议放在插件根的 `.claude-plugin/plugin.json`，然后用 README 中的完整命令做一次端到端安装验证。

File: .agents/skills/scaffolding/SKILL.md [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+    ),
+)
+```
+
+See [rpc/triple/openapi](https://github.com/apache/dubbo-go-samples/tree/main/rpc/triple/openapi).
+
+**Mount an HTTP handler** on the same Triple port (Triple-only):
+
+```go
+srv.AttachHTTPHandler("/healthz", http.HandlerFunc(healthz))
```
Comment (AlexStocks) [inline .agents/skills/scaffolding/SKILL.md:226]: [P1] 生成的 HTTP handler 示例无法编译
当前 `server.(*Server).AttachHTTPHandler` 的签名只有一个参数：`AttachHTTPHandler(handler http.Handler) error`，不接受 path。把本示例放进独立消费者模块编译会报 `too many arguments ... have (string, http.HandlerFunc), want (http.Handler)`。同样的错误还出现在 `guide/SKILL.md` 和 `migrate/SKILL.md`。这与 README 宣称“生成可直接编译的骨架”冲突。建议先用 `http.NewServeMux()` 在 mux 上注册 `/healthz`、`/api` 等路径，再把整个 mux 作为唯一参数传给 `AttachHTTPHandler`，并给所有 Go 示例增加自动编译测试。

File: .agents/skills/extensions/SKILL.md [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+func init() {
+    extension.SetLoadbalance("tenant-affinity", func() loadbalance.LoadBalance {
+        return &tenantAffinityLB{}
+    })
+}
+
+type tenantAffinityLB struct{}
+
+func (lb *tenantAffinityLB) Select(invokers []base.Invoker, inv base.Invocation) base.Invoker {
+    tenant := inv.Attachment("tenant-id")
```
Comment (AlexStocks) [inline .agents/skills/extensions/SKILL.md:140]: [P1] 示例调用了不存在的 Invocation API
`protocol/base.Invocation` 只提供 `GetAttachment`、`GetAttachmentInterface`、`GetAttachmentWithDefaultValue` 等方法，没有 `Attachment`。独立消费者探针按这里的代码编译会报 `inv.Attachment undefined`；同一错误还出现在本文件前面的附件说明。建议根据需要改为 `GetAttachment` 或 `GetAttachmentInterface`，并把 skill 中的代码块纳入编译门禁，否则 Agent 会稳定生成无法构建的扩展实现。
